docs/demo/data/processed/model_extensions.py

```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Dec 19 10:50:55 2023
+
+@author: bdobson
+"""
+from wsimod.core import constants
+import sys
+import os
+from tqdm import tqdm
+from math import log10
+from wsimod.nodes.land import ImperviousSurface
+from wsimod.nodes.nodes import QueueTank, Tank, ResidenceTank
+
+def extensions(model):
+    # Apply customations
+    model.nodes['my_groundwater'].residence_time *= 2
+    
+    # Add dcorator to a reservoir
+    model.nodes['my_reservoir'].net_evaporation = model.nodes['my_reservoir'].empty_vqip()
+    model.nodes['my_reservoir'].mass_balance_out.append(lambda self=model.nodes['my_reservoir']: self.net_evaporation)
+    model.nodes['my_reservoir'].net_precipitation = model.nodes['my_reservoir'].empty_vqip()
+    model.nodes['my_reservoir'].mass_balance_in.append(lambda self=model.nodes['my_reservoir']: self.net_precipitation)
+
+    model.nodes['my_reservoir'].make_abstractions = wrapper_reservoir(model.nodes['my_reservoir'], model.nodes['my_reservoir'].make_abstractions)
```

Comment:
Is there any reason why this type of extensions cannot be implemented via a custom node? Something like:

```python
# Inherit from the relevant Node sublcass, of course
class MyReservoir(Node):
    def __init__(self, ...):
        super().__init__(self, ...)
        self.net_evaporation = self.empty_vqip()
        self.net_precipitation = self.empty_vqip()
        self.mass_balance_out.append(self.net_evaporation)
        self.mass_balance_in.append(self.net_precipitation)

    def make_abstractions(self):
        # do custom stuff to the node
        ...
        return super().make_abstractions()
```

And then use this new node in the input file.

While your approach works, it requires setting two different ways of adding extra functionality and while, as discussed, we do want that flexibility, it will be much better if those additions can be somewhat streamlined into a few specific approaches. Creating custom Nodes seems like the best approach as thy are used in very specific places and very specific ways. In combination with the workflow customisation added by @barneydobson , that's a lot of fleixbility.

Reply:
To me, the reason why we have two different ways of customisation is:
(1) sometimes we want the node keeping the default designs, we just want to change the values of some parameters.
    This is essential, as we plan to develop wrappers to enable sensitivity tests in the future, which needs the portal to only change the parameters.
(2) sometimes we want the new node that has a new design (functions and attributes), we need to write a new class.

Formulating a unified way to enable these two situations is definitely great. But to me I feel like keep them separate is more logical, which may not be the most user-friendly way.

Reply:
> (1) sometimes we want the node keeping the default designs, we just want to change the values of some parameters. This is essential, as we plan to develop wrappers to enable sensitivity tests in the future, which needs the portal to only change the parameters.

Cannot we just define this override in the input file when requesting a node? Maybe in an `override` section? If it is just a value, I see no reason why you cannot do sensitivity analysis with this.

In my opinion, if some user interface is not user friendly, then the underlying architecture resulting in that interface needs improvement. As developer, we can modify the software as much as we can if that makes life for end users easier. In this particular case, if customisation can be done via de config file for simple cases and custom nodes for the more complicated ones, that will make the process more robust and reproducible, then that's a step we should take. Enabling a free for all manipulation of the model can only end in disaster...

Reply:
@barneydobson , any comment on this?

Reply:
You are definitely correct @dalonsoa. For sensitivity test, I've made new changes on the code to let the model receive a dict when loading the model. This new dict contains the information of changes in the yaml and will be used to replace the original yaml dict before adding nodes/arcs. It works as expected and I've push the changes to a new channel 'enable_receiving_parameters'.

However, the key challenge here is that most nodes have parameters that are initialised automatically and do not need external args input. These parameters cannot be changed via overriding the yaml dict but can only be changed after the model has been built. One way for unifying this is to substantially rewrite all the node scripts to compulsorily initialise every parameter via args input - then all the parameters can be only changed via overriding. I'll let @barneydobson to comment on this.

Reply:
I don't think it will be necessary to re-write everything, but just to check what parameters are in the `override` section. If they are there, you use them, otherwise, you just use the automatically initialised versions. 

Having said that, I have not check the part of the code that does that, so it might be that things are more complicated than I thought.

Reply:
Sorry for the delay on this - I've had a sick baby last week then been sick myself - I'll hope to look over this soonest

Reply:
sick babies do take precedence! So no worry at all - specially after all the time it took me to go back to this.

Reply:
> sick babies do take precedence! So no worry at all - specially after all the time it took me to go back to this.

It's OK, I am the sick baby now. Gently going over GH issues I'm sure is the cure!

Reply:
Right - but I think we would need to generalise this beyond the `NutrientPool` and conceive that a user may want to change any attribute in the model. So I guess based on this the two options are:
- Setters that enable a function to be called when certain parameters are set. In the case of extensions, this means that the overrides are applied after the model/nodes/arcs are initialised.
- Expand argument passing so that everything can ultimately be set when initialising the parent node/object. In the case of extensions, this means that the overrides are applied before the model/nodes/arcs are initialised. I also think this means there wouldn't be an `overrides` section in the config file as I showed in my example above, since if the parameters are default attributes to an object they will get saved during `model.save()`

I see pros for both.

The 'setters' case 1) it seems more tidy to program, avoiding a huge amount of argument passing - possibly slightly more involved to implement but ultimately fewer lines of code, 2) it feels like it enables more robust use of the model (e.g., if someone changes one of these parameters halfway through a simulation, it has the best chance of doing something sensible). 3) puts less onus on the user to understand complicated things, e.g., they don't have to be aware that - in setting one attribute, you would need to change values of another

The 'arguments' case 1) seems conceptually easier to grasp? though the more I think about it the less convinced I am that this is actually the case 2) in the `config` file all of my model parameters are in the places 'that they live' (that said, with setters you could achieve this too, i.e., if the parameter is an `init` argument then pass it directly, otherwise `init` the object and then call the setter - though I can think of edge cases where doing the overrides after _all_ of the model nodes/arcs have been initialised is probably more sensible). 3) less complicated code to understand, arguments are easy.

@dalonsoa I think your read on this would be helpful if possible! Though as I've revised this I feel like the first case is probably the better option - the pros in the second case aren't convincing me. @liuly12 if you wanted to add any pros/cons please go ahead

Reply:
Ok, give me a couple of days to digest this and truly understand the different use cases.

Reply:
RE your example - in that case we would have to write a setter function for the `total_porosity` that meaningfully effects these changes? The setter could even reinitialise the tank. I think the benefit of setters is that they mean we only need to implemet changes where they are needed, rather than across all nodes

Reply:
Okay, now I get it. The 'setter' can save efforts in revising the code in the nodes, but it probably needs to extensively pick up all the potential nodes that should be reinitialised in this way. It might be worth thinking whether the reinitialisation of all these nodes follow exactly the same way or there are differences.

Reply:
Agree - the effort will not be insubstantial! I wonder if we create an issue for each node. @dalonsoa once you've digested this it'd be good to get an idea what is a standard practice towards something like this

Reply:
Ok, lets see if I got this right. The nodes are all initialised (unless created programmatically) in this line:

https://github.com/ImperialCollegeLondon/wsi/blob/e3b66fefa423f93bb698f4aa16c655e7010a0c87/wsimod/orchestration/model.py#L418-L418

Where the `data` is all the configuration associated to the node of interest in the config file, right?

The doubt to enable an easier customisation of the nodes by the user are:

1. expand the constructor to accept more input parameters and include those in the config file, so the node is created with the right parameters at initialisation time,
2. leave the constructor as it, creating a fully working node with some default parameters, and then update selected parameters included in an override section of the config file via setters such that any side effects of those changes are propagated to the right places.

Is this correct?

I would go for the second option, which has also the advantage that is backwards compatible. I think it makes for a cleaner interface to have a core configuration that will produce a working model and then "tweak" the models based on overrides. I would do something like the following after adding the nodes and the arcs in:

- [run_model](https://github.com/ImperialCollegeLondon/wsi/blob/e3b66fefa423f93bb698f4aa16c655e7010a0c87/wsimod/__main__.py#L45)
- [model.load](https://github.com/ImperialCollegeLondon/wsi/blob/e3b66fefa423f93bb698f4aa16c655e7010a0c87/wsimod/orchestration/model.py#L152)

```python
# Initialise nodes
model.add_overrides(data.get("overrides", {}))

# And in Model
def add_overrides(self, config: Dict):
    for node in config.get("nodes", {}):
        type_ = node.pop("type_")
        name = node.pop("name")
        self.nodes_type[type_][name].apply_overrides(node)

    for arc in config.get("arcs", {}):
        name = arc.pop("name")
        self.arcs[name].apply_overrides(arc)
``` 

The `node.apply_overrides` and `arc.apply_overrides` will be the methods that will call the appropriate setters to ensure that any side effects of modifying those parameters are taken care of. 

`apply_overrides` should be a method (not an abstract one) of the base class for `nodes` and `arcs` that will do nothing unless it is overridden. This way, only those specific nodes that can be overridden will need to have this method (and the relevant setters) implemented. 

```python
class Node:
    def apply_overrides(self, config: Dict):
        pass

class Arc:
    def apply_overrides(self, config: Dict):
        pass
```

In terms of how this can be done from a practical point of view, I would open issues for:

1. Implement the above overriden logic. Should be harmless and backwards compatible with any existing code and config files.
2. Identify which nodes/arcs have parameters that are worth to be overriden and open an issue for each of those nodes/arcs, indicating which parameters need such override. Watch out for recursive calls! You might need some hidden internal attributes that can be set directly and only use the setters in the overrides:

```python
class MyNode:
    def __init__(...):
        # Does not use the setter
        self._some_param = get_default_some_param(...)

    @property 
    def some_param(self):
        return self._some_param

    @some_param.setter 
    def some_param(self, value):
        if not value:
            return
        # Some complex logic
        self._some_param = ...
        # More logic here

    def apply_overrides(self, config: Dict):
        self.some_param = config.pop("some_param", None)
        ...
```

3. Implement the above issues, **one PR per issue**, no more!

There might be some extra things to do, but I would say it is a safe approach, backwards compatible and that should give plenty of flexibility to pretty much do whatever you want via the configuration file. For the most complex cases, naturally the user can create their custom nodes or arcs, which I think we have agreed is useful to have.

Does this make sense?

Reply:
Yep - I will do the high level changes in `Node` and `Model`, and then do an example in `WWTW`. After this I think your focus should be on `Land`, since that will by far the most complicated - could even do an issue per `Surface` - start with the base surface and we'll see how many changes there are on the PR.

Reply:
See issues:
#66 
#67 
#68